server/src/instant/db/model/attr.clj

```diff
@@ -257,6 +257,25 @@
                      "Namespaces are not allowed to start with a `$`.
                       Those are reserved for system namespaces.")}])))))
 
+(defn validate-add-required! [conn app-id attrs]
+  (doseq [attr  attrs
+          :when (:required? attr)
+          :let  [[_ etype label] (:forward-identity attr)
+                 query "SELECT *
+                          FROM idents
+                          JOIN attrs ON attrs.forward_ident = idents.id
+                          JOIN triples ON triples.attr_id = attrs.id
+                         WHERE idents.etype = ?etype
+                           AND idents.app_id = ?app-id
+                           AND triples.app_id = ?app-id
```

Comment:
You may be able to make this faster by writing `AND ea` (as all of these triples will live on that index)

Reply:
Not necessarily. What I’m checking here is that type has at least 1 entity (thus `LIMIT 1`). But if it only has entity with cardinality many it won’t be on `ea`?

Also, because of LIMIT 1, we don’t really need index here, it’s a simple existence check

Reply:
Ah, I originally assumed required would only work on attributes, but I realize it can work for links too! Good point. 

Perhaps based on the attr we could infer if it's cardinality-one, and use ea, or value-type ref, and use eav. I am not sure if it would help though.